src/Storages/ObjectStorage/DataLakes/Iceberg/IcebergMetadata.cpp

```diff
@@ -778,34 +833,34 @@ std::optional<size_t> IcebergMetadata::totalBytes(ContextPtr local_context) cons
 ObjectIterator IcebergMetadata::iterate(
     const ActionsDAG * filter_dag,
     FileProgressCallback callback,
-     size_t /* list_batch_size */,
-     ContextPtr local_context) const
+    size_t /* list_batch_size */,
+    StorageSnapshotPtr storage_snapshot,
+    ContextPtr local_context) const
 {
-    SharedLockGuard lock(mutex);
+    auto iceberg_table_state = extractIcebergSnapshotIdFromMetadataObject(storage_snapshot);
+    if (iceberg_table_state == nullptr)
+    {
+        throw Exception(
+            ErrorCodes::LOGICAL_ERROR,
+            "Can't extract iceberg table state from storage snapshot for table location {}",
+            persistent_components.table_location);
+    }
+    chassert(iceberg_table_state != nullptr);
```

Comment:
This assert is redundant as right before that you check `if (iceberg_table_state == nullptr)` and throw an error in this case